### PR TITLE
Node launcher

### PR DESCRIPTION
**Pull request by kruplm**
Story: "As a frontend developer I can configure the height, width, title and the URL of a browser window from my JavaScript application."

- the shell can be launched from Node.JS as well.
- width, height and title of the window can be specified.
- An url can be passed via configuration or by the LoadUrl() function
- The default config loading (with pre-existing manifest file) still works the same way - so that our future Universal hosting goals can be achieved.
- As part of this PR the manifest file parsing was also improved.

### CODE REVIEW

File: Tryouts/Prototypes/Shell/Node-launcher/Lib/src/Launcher.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                this.argsArray.push(`--height=${_config?.height}`);
+            }
+        }
+
+        console.debug("\t[]", this.argsArray);
+    }
+
+    
+    public launch() {
+        //TODO replace after application is properly packaged and added to PATH
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/Node-launcher/Lib/src/Launcher.ts:35]: Can we remove these early on? Just add the directory to your PATH variable (and document it).

File: Tryouts/Prototypes/Shell/App.xaml.cs [hunk 2]
```diff
@@ -28,6 +28,11 @@ namespace Shell
     /// </summary>
     public partial class App : Application
     {
-        
+        public string[] CustomerCode { get; set; }
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/App.xaml.cs:31]: What customer code? Who's the customer and what kind of code is provided here? It's actually populated from command line arguments, so why isn't it called `CommandLineArguments`?

File: Tryouts/Prototypes/Shell/MainWindow.xaml.cs [hunk 3]
```diff
@@ -40,13 +41,59 @@ public partial class MainWindow : Window
...
+            if (commands.ContainsKey("url"))
+            {
+                this.commandLineURL = commands["url"];
+            }
+            else 
+            {
+                this.commandLineURL = "about:blank";
+            }
+
+            var lazyWebContent = new WebContent(this.commandLineURL);
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWindow.xaml.cs:87]: Don't call it lazy if it's not a `Lazy<T>` or similar.

File: Tryouts/Prototypes/Shell/MainWindow.xaml.cs [hunk 4]
```diff
@@ -40,13 +41,59 @@ public partial class MainWindow : Window
...
+        }
+
+        private void LazyLoading() {
+            Dictionary<string, string> commands = new Dictionary<string, string>();
+
+            this.hereIsMyCode.ToList<string>().ForEach(item => {
+                item = item.TrimStart('-');
+                string[] command = item.Split("=");
+
+                commands.Add(command[0], command[1]);
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWindow.xaml.cs:75]: This will add the wrong url if it contains the '=' character, which is common.

File: Tryouts/Prototypes/Shell/MainWindow.xaml.cs [hunk 5]
```diff
@@ -40,13 +41,59 @@ public partial class MainWindow : Window
...
+            });
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWindow.xaml.cs:76]: Don't invent the wheel, use the `System.CommandLine` library instead, it's well tested, robust, and pretty standard.

File: Tryouts/Prototypes/Shell/MainWindow.xaml.cs [hunk 6]
```diff
@@ -40,13 +41,59 @@ public partial class MainWindow : Window
...
+
+            if (commands.ContainsKey("url"))
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWindow.xaml.cs:78]:
```cs
if (commands.TryGetValue("url", out var url) 
{
    this.commandLineURL = url;
}
```
or
```
this.commandLineURL = commands.GetValueOrDefault("url");
```
etc. 
But you should just ditch the whole thing and use `System.CommandLine` or similar parsing library.

File: Tryouts/Prototypes/Shell/MainWindow.xaml.cs [hunk 7]
```diff
@@ -40,13 +41,59 @@ public partial class MainWindow : Window
         private ManifestModel config;
         private ModuleModel[]? modules;
 
+        private string commandLineURL;
+        private string[] hereIsMyCode = (Application.Current as App).CustomerCode;
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWindow.xaml.cs:45]: Don't invent names. Rename to `commandLineArgs` or something that says what it does.

File: Tryouts/Prototypes/Shell/MainWindow.xaml.cs [hunk 8]
```diff
@@ -40,13 +41,59 @@ public partial class MainWindow : Window
...
 
-            config = new ManifestParser().Manifest;
-            modules = config.Modules;
-            DataContext = modules;
+            this.Loaded += MainWindow_Loaded;
+        }
+
+        private void MainWindow_Loaded(object sender, RoutedEventArgs e)
+        {
+            if (this.hereIsMyCode.Length != 0)
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWindow.xaml.cs:56]: What's the expected behavior here? If I provide the url, an empty `MainWindow` will pop up. Consider hiding the main window in this case.

File: Tryouts/Prototypes/Shell/Node-launcher/Lib/rollup.config.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿import { nodeResolve } from '@rollup/plugin-node-resolve';
+import typescript from '@rollup/plugin-typescript';
+
+export default {
+    input: 'src/index.ts',
+    output: {
+      
+        file: 'output/bundle.mjs',
+        format: 'es',
+        name: "@compose/node-launcher"
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/Node-launcher/Lib/rollup.config.js:10]: Why is the scope and name different from the one in `package.json`? Is this something rollup-specific?

File: Tryouts/Prototypes/Shell/Node-launcher/Lib/src/BrowserWindow.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
+export class BrowserWindow {
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/Node-launcher/Lib/src/BrowserWindow.ts:1]: What's this class for? Is it a stub for future implementation?

File: Tryouts/Prototypes/Shell/MainWindow.xaml.cs (hunk 7 above)
Comment (ztanczos) [inline Tryouts/Prototypes/Shell/MainWindow.xaml.cs:45]: If you know the type of Application.Current is always App then use direct cast. If you don't know/can't guarantee that then using `as` is correct, however you should check for null result then.

File: Tryouts/Prototypes/Shell/Manifest/ManifestParser.cs [hunk 11]
```diff
@@ -29,7 +30,13 @@ internal class ManifestParser
...
+            string fileName = "exampleManifest.json";
+            string processPath = Environment.ProcessPath;
+            string folder= Path.GetDirectoryName(processPath);
+
+            string path = Path.Combine(folder, @"Manifest\", fileName);
+            
+            OpenManifestFile(path);
         }
 
         public async void OpenManifestFile(string manifestFile)
```
[the remark this answers is not in the document]
Comment (ztanczos) [reply Tryouts/Prototypes/Shell/Manifest/ManifestParser.cs:42]: +1, but also don't use `async void` methods, except for evend handlers. See https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming for details.

[the remark this answers is not in the document]
Comment (BalassaMarton) [reply Tryouts/Prototypes/Shell/Manifest/ManifestParser.cs:42]: In defense of `async void`, if it's clear that it won't be awaited and you handle exceptions inside the method, it should be fine.

File: Tryouts/Prototypes/Shell/Node-launcher/Lib/src/BrowserWindow.ts (hunk 10 above)
Comment (kruplm) [reply Tryouts/Prototypes/Shell/Node-launcher/Lib/src/BrowserWindow.ts]: Yes, I accidentally committed that WIP code after I rearranged the project files for the monorepo.
I've pushed the intended code just now.

File: Tryouts/Prototypes/Shell/Node-launcher/Lib/rollup.config.js (hunk 9 above)
Comment (kruplm) [reply Tryouts/Prototypes/Shell/Node-launcher/Lib/rollup.config.js:10]: It's the global variable name that represents the bundle if it's iife or umd, But it's not needed for ES so I'm going to remove it.

File: Tryouts/Prototypes/Shell/MainWindow.xaml.cs (hunk 6 above)
Comment (kruplm) [reply Tryouts/Prototypes/Shell/MainWindow.xaml.cs]: First time I checked System.CommandLine, it was still in preview. Other parsing libraries have licenses that aren't good for this project, or hasn't been maintained for a while.
My idea was to switch to System.CommandLine, and rather deal with it's changes in the future. But I wanted to discuss it with the team first.

Comment (ztanczos) [reply Tryouts/Prototypes/Shell/MainWindow.xaml.cs]: We should be good to use System.CommandLine, that package is already being used by large projects (e.g.: dotnet CLI itself).

Comment (kruplm) [reply Tryouts/Prototypes/Shell/MainWindow.xaml.cs]: That sounds reassuring.

File: Tryouts/Prototypes/Shell/Manifest/ManifestParser.cs [hunk 12]
```diff
@@ -20,26 +20,28 @@
...
 using System.Windows.Controls.Primitives;
+using System.Windows.Documents;
 
 namespace Manifest
 {
-    internal class ManifestParser
+    internal static class  ManifestParser
     {
-        internal ManifestModel Manifest { get; set; }
+        internal static ManifestModel Manifest { get; set; }
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/Manifest/ManifestParser.cs:29]: Unnecessary property. Why would you have a static parser that also stores the last manifest it has parsed?

File: Tryouts/Prototypes/Shell/Manifest/ManifestParser.cs [hunk 13]
```diff
@@ -20,26 +20,28 @@
...
 
-        public ManifestParser()
+        public static  ManifestModel OpenManifestFile(string fileName)
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/Manifest/ManifestParser.cs:31]: `OpenManifestFile` suggests it returns something that can be later closed. Just call it `Parse` or `ParseManifest`.

File: Tryouts/Prototypes/Shell/Manifest/ManifestParser.cs [hunk 14]
```diff
@@ -20,26 +20,28 @@
...
-        }
+            string processPath = Environment.ProcessPath;
+            string folder = Path.GetDirectoryName(processPath);
+            string path = Path.Combine(folder, @"Manifest\", fileName);
 
-        public async void OpenManifestFile(string manifestFile)
-        {
-             using (FileStream stream = File.Open(manifestFile, FileMode.Open))
-             {
+            using (FileStream stream = File.Open(path, FileMode.Open))
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/Manifest/ManifestParser.cs:37]: Suggestion:
```cs
using (var stream = File.OpenRead(path))
```
Edge case, but the original version will fail if you start two instances of the program quickly, because it tries to open the file for read/write access and without sharing.

File: Tryouts/Prototypes/Shell/Node-launcher/Lib/src/Launcher.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+import { WindowConfig } from './WindowConfig';
+
+export class Launcher {
+    private argsArray: string[] = [];
+
+    constructor(_config?: WindowConfig ) { 
+
+    }
+
+    private processArgs(_config?: WindowConfig) {
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/Node-launcher/Lib/src/Launcher.ts:11]: I'm confused about naming conventions here. Why the underscore before the function parameter?

File: Tryouts/Prototypes/Shell/Node-launcher/Lib/src/WindowConfig.ts [hunk 16]
```diff
@@ -0,0 +1,5 @@
+export interface WindowConfig {
+    url: string;
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/Node-launcher/Lib/src/WindowConfig.ts:2]: Is this config object all or nothing? All these properties seem to be optional (and they are all treated as optional).

File: Tryouts/Prototypes/Shell/Properties/launchSettings.json [hunk 17]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/Properties/launchSettings.json:1]: Is this file really necessary?

**Review by BalassaMarton — CHANGES_REQUESTED** [review]

File: Tryouts/Prototypes/Shell/MainWebWindow.xaml.cs [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+using System.Windows.Shapes;
+
+namespace Shell
+{
+    /// <summary>
+    /// Interaction logic for MainWebWindow.xaml
+    /// </summary>
+    public partial class MainWebWindow : Window
+    {
+        private string[] commandLineArguments = (Application.Current as App).CommandLineArguments;
```
[the remark this answers is not in the document]
Comment (ztanczos) [reply Tryouts/Prototypes/Shell/MainWebWindow.xaml.cs:22]: Another alternative: command line arguments could be passed to the objects which need them via the ctor.

File: Tryouts/Prototypes/Shell/MainWindow.xaml.cs [hunk 19]
```diff
@@ -71,7 +72,7 @@ private void ShowChild_Click(object sender, RoutedEventArgs e)
         {
             var context = (sender as Button).DataContext;
```
Comment (ztanczos) [inline Tryouts/Prototypes/Shell/MainWindow.xaml.cs:73]: In this case the sender's `DataContext` is the same as the current Window's `DataContext`. But an even better approach would be to store the DataContext in a typed private field so that you wouldn't need to cast.

File: Tryouts/Prototypes/Shell/App.xaml [hunk 20]
```diff
@@ -8,12 +8,13 @@ is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND
 See the License for the specific language governing permissions and limitations under the License.
 ********************************************************************************************************
 -->
-
+<!--StartupUri="MainWindow.xaml"-->
```
Comment (ztanczos) [inline Tryouts/Prototypes/Shell/App.xaml:11]: Please remove commented out code

File: Tryouts/Prototypes/Shell/Node-launcher/Lib/src/WindowConfig.ts (hunk 16 above)
Comment (kruplm) [reply Tryouts/Prototypes/Shell/Node-launcher/Lib/src/WindowConfig.ts]: The parameters are optional because the user might want to set the url, but not the width/height. And vice versa.

File: Tryouts/Prototypes/Shell/App.xaml.cs [hunk 21]
```diff
@@ -28,6 +28,23 @@ namespace Shell
     /// </summary>
     public partial class App : Application
     {
-        
+        public string[] CommandLineArguments { get; set; }
```
Comment (ztanczos) [inline Tryouts/Prototypes/Shell/App.xaml.cs:31]: This property is not used anymore, you can remove it.

File: Tryouts/Prototypes/Shell/MainWebWindow.xaml.cs [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+using System.Security.Policy;
+
+namespace Shell
+{
+    /// <summary>
+    /// Interaction logic for MainWebWindow.xaml
+    /// </summary>
+    public partial class MainWebWindow : Window
+    {
+        public string Url;
```
Comment (ztanczos) [inline Tryouts/Prototypes/Shell/MainWebWindow.xaml.cs:26]: Unused public field - you can remove it.

File: Tryouts/Prototypes/Shell/CommandLineArgumentsParser.cs [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+
+        public CommandLineArgumentsParser(string[] _commandLineArguments)
+        {
+            commandLineArguments = _commandLineArguments;  
+        }
+        public async void CreateDynamicWindowWithOptions()
+        {
+            var titleOption = new Option<string>("--title", description: "Set title for window");
+            var urlOption = new Option<string>("--url", description: "Set url for webview. default: about:blank");
+            var widthOption = new Option<int>("--width", description: "Set width for window");
```
Comment (ztanczos) [inline Tryouts/Prototypes/Shell/CommandLineArgumentsParser.cs:23]: Do we have any defaults for width/height?

File: Tryouts/Prototypes/Shell/CommandLineArgumentsParser.cs [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+{
+    internal class CommandLineArgumentsParser
+    {
+        private string[] commandLineArguments;
+
+        public CommandLineArgumentsParser(string[] _commandLineArguments)
+        {
+            commandLineArguments = _commandLineArguments;  
+        }
+        public async void CreateDynamicWindowWithOptions()
```
Comment (ztanczos) [inline Tryouts/Prototypes/Shell/CommandLineArgumentsParser.cs:19]: I'd change this to return a Task instead of async void to improve testability.

File: Tryouts/Prototypes/Shell/CommandLineArgumentsParser.cs [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            var titleOption = new Option<string>("--title", description: "Set title for window");
+            var urlOption = new Option<string>("--url", description: "Set url for webview. default: about:blank");
+            var widthOption = new Option<int>("--width", description: "Set width for window");
+            var heightOption = new Option<int>("--height", description: "Set height for window");
+
+            var rootCommand = new RootCommand();
+
+            rootCommand.Add(titleOption);
```
Comment (ztanczos) [inline Tryouts/Prototypes/Shell/CommandLineArgumentsParser.cs:28]: Nitpick: you can use the collection initializer syntax here:
```
var rootCommand = new RootCommand
{
  urlOption,
  widthOption,
...
};
```

File: Tryouts/Prototypes/Shell/MainWebWindowOptions.cs [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Shell
+{
+    public sealed class MainWebWindowOptions
+    {
+        public string Title { get; set; }
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWebWindowOptions.cs:11]: If any of these properties are optional, they should have nullable type like `string?` or `int?`

File: Tryouts/Prototypes/Shell/MainWebWindowOptionsParser.cs [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+        }
+
+        public MainWebWindowOptions Parse(string[] args)
+        {
+            Parser parser = new Parser(rootCommand);
+            ParseResult parseResult = parser.Parse(args);
+
+            MainWebWindowOptions options = new MainWebWindowOptions
+            {
+                Title = parseResult.GetValueForOption(titleOption) ?? "Compose Web Container",
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWebWindowOptionsParser.cs:36]: Some tests would be useful, I'm not sure what value is returned when the option is not set (it might even be an exception that we probably don't want)

File: Tryouts/Prototypes/Shell/MainWebWindowOptionsParser.cs [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+using System.CommandLine.Parsing;
+using System.Linq;
+using System.Security.Policy;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows.Media.Media3D;
+
+namespace Shell
+{
+    internal class MainWebWindowOptionsParser
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWebWindowOptionsParser.cs:14]: This class should be static.

File: Tryouts/Prototypes/Shell/CommandLineArgumentsParser.cs (hunk 25 above)
Comment (kruplm) [reply Tryouts/Prototypes/Shell/CommandLineArgumentsParser.cs]: The nitpick helped to turn the class to static, I needed to have a constructor before for this.

File: Tryouts/Prototypes/Shell/MainWebWindowOptions.cs (hunk 26 above)
Comment (kruplm) [reply Tryouts/Prototypes/Shell/MainWebWindowOptions.cs]: since Window.Width and Height are double (fixed the wrong type now) instead of double? I can't set it double? here because I'd need to cast where I assign it.  These are never null because the default value is 0.0. 
And if the user doesn't specify the value from the JavaScript API the window is set to 450x800 now.

**Review by BalassaMarton — APPROVED** [review]

File: Tryouts/Prototypes/Shell/MainWebWindowOptions.cs [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+namespace Shell
+{
+    public sealed class MainWebWindowOptions
+    {
+        public double? Height { get; set; }
+        public string? Title { get; set; }
+        public string? Url { get; set; }
+        public double? Width { get; set; }
+
+        public double DefaultHeight { get; set; } = 450d;
```
Comment (BalassaMarton) [inline Tryouts/Prototypes/Shell/MainWebWindowOptions.cs:16]: These should be static constants with non-nullable type (`public const double DefaultHeight = 450` etc.)

**Review by BalassaMarton — CHANGES_REQUESTED** [review]

**Review by BalassaMarton — APPROVED** [review]
